File: my_problems/day50-dsa-hashing-I/hw_q3_isPalindrome.java
import java.util.HashMap;
import java.util.Map;

public class hw_q3_isPalindrome {

    public static int checkPalindrome(String A) {

        // the palindrome string can be formed if frequency of every character in string
        // is even except single character which can have odd frequency

        Map<Character, Integer> map = new HashMap<Character, Integer>();
        for (int i = 0; i < A.length(); i++) {
            char currentChar = A.charAt(i);
            if (map.containsKey(currentChar)) {
                int currentFreq = map.get(currentChar);
                map.put(currentChar, currentFreq + 1);
            } else {
                map.put(currentChar, 1);
            }
        }

        // if off count is >1 , the palindrome can not be possible
        int countOfOdd = 0;
        for (Character key : map.keySet()) {
            int freq = map.get(key);
            if (freq % 2 == 1) {
                countOfOdd++;
            }
            if (countOfOdd > 1) {
                // can not be palindrome
                return 0;
            }
        }

        // palindrome can be possible
        return 1;
    }

    public static void main(String[] args) {

        int output1 = checkPalindrome("abcde");
        System.out.println("answer -> " + output1);
        int output2 = checkPalindrome("abbaee");
        System.out.println("answer -> " + output2);
    }
}

File: my_problems/day50-dsa-hashing-I/hw_q3_isPalindrome.js
/* Check Palindrome - II */

const checkPalindrome = function (A) {

    // To be valid palindrome, each character should occur in multiple of 2 except one in middle
    // Approach: find frequency iof each character within string
    const array = A.split("");

    // build hashMap of <Character, frequency>
    // hashMap<String, Integer> => <Character, frequency>
    let hashMap = new Map();
    for (let i = 0; i < array.length; i++) {
        if (hashMap.has(array[i])) {
            let currentFrequency = hashMap.get(array[i]);
            hashMap.set(array[i], currentFrequency + 1);
        } else {
            hashMap.set(array[i], 1);
        }
    }

    //  only single odd frequency is allowed
    let oddFrequencyCount = 0
    const keys = hashMap.keys();
    for (let key of keys) {
        // odd frequency
        if ((hashMap.get(key) & 1) === 1) {
            oddFrequencyCount++;
            if (oddFrequencyCount > 1) {
                break;
            }
        }
    }

    if (oddFrequencyCount > 1) {
        return 0;
    }
    return 1;
}

console.log(checkPalindrome("abcde"));
console.log(checkPalindrome("abbaee"));